src/Core/Lib/BTreeDictionary.cs

```diff
@@ -0,0 +1,750 @@
+                right.count = count - iSplit;
+                this.count = iSplit;
+                right.totalCount = right.count;
+                this.totalCount = this.count;
+                Array.Copy(this.keys, iSplit, right.keys, 0, right.count);
+                Array.Clear(this.keys, iSplit, right.count);
+                Array.Copy(this.values, iSplit, right.values, 0, right.count);
+                Array.Clear(this.values, iSplit, right.count);
+                right.nextLeaf = this.nextLeaf;
+                this.nextLeaf = right;
+                if (tree.Comparer.Compare(right.keys[0], key) < 0)
+                    right.Add(key, value, tree);
+                else
+                    this.Add(key, value, tree);
+                return right;
+            }
+        }
+
+        /// <summary>
+        /// In a B+tree, the internals node only contain links to other nodes.
+        /// </summary>
+        private class InternalNode : Node
+        {
+            public Node[] nodes;
+
+            public InternalNode(int children)
+            {
+                this.keys = new TKey[children];
+                this.nodes = new Node[children];
+            }
+
+            public override (Node, Node) Add(TKey key, TValue value, BTreeDictionary<TKey, TValue> tree)
+            {
+                int idx = Array.BinarySearch(keys, 1, count-1, key, tree.Comparer);
+                int iPos;
+                if (idx >= 0)
+                    iPos = idx - 1;
+                else
+                    iPos = (~idx) - 1;
+                var subnode = nodes[iPos];
+                var (leftNode, rightNode) = subnode.Add(key, value, tree);
+                if (rightNode == null)
+                {
+                    this.totalCount = SumNodeCounts(this.nodes, this.count);
+                    return (leftNode, null);
+                }
+                return Insert(iPos + 1, rightNode.keys[0], rightNode,  tree);
+            }
+
+            public Node Add(TKey key, Node node, BTreeDictionary<TKey, TValue> tree)
+            {
+                int idx = Array.BinarySearch(keys, 1, count-1, key, tree.Comparer);
+                if (idx >= 0)
+                    throw new ArgumentException("Duplicate key.");
+                var subnode = nodes[~idx];
+                return Insert(~idx, node.keys[0], node, tree).Item1;
+            }
+
+            public override (TValue, bool) Get(TKey key, BTreeDictionary<TKey, TValue> tree)
+            {
+                int idx = Array.BinarySearch(keys, 1, count - 1, key, tree.Comparer);
+                if (idx >= 0)
+                    return nodes[idx].Get(key, tree);
+                else
+                {
+                    var iPos = (~idx) - 1;
+                    return nodes[iPos].Get(key, tree);
+                }
+            }
+
+            public override (Node, Node) Set(TKey key, TValue value, BTreeDictionary<TKey, TValue> tree)
```

Comment:
`Set` is almost the same as `Add`. Maybe they are should be merged to one method with extra `bool` flag